# Free energy {#adjMEF}

We have five collections of 5'UTR and use the `ShortRead`, `BioStrings` and `BSgenome.Hsapiens.UCSC.hg38` packages to get the squences.

1. All annotated 5' UTR (79,176): transcript-based and unique 5' UTR model from Gencode v35
2. DUX4 targets: 88 targets identified from Polysome-seq, and the 5'UTR include annotated and un-annotated identified by the `SESeq` packages; `DUX4-targets-polysome-84` how did I get this? Hint: `110-HDUX4-promoter.R` checked if the candidate has alternative splice events, alternative first exon (AFE) or alternative start site (AS). Some genes use alernative start site / first exon, but some dont'. `DUX4-targets-novel-and-annotated-5'UTR.xlsx`.
3.
4. Down-regulated translational efficiency by Ribo-seq (S6) in 5'UTR region: `TE down Ribo-seq`
5. TE list from poly-seq: `TE dwon Poly-seq`, total 5,800 genes; high/sub


The plan to calculate the free energy of the five collections of 5'UTR is implemented by `09B-5UTR-structure.R`. 