<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 7 5’ UTR Free Energy Evaluation | DUX4 ribosome footprints profiling and translation efficiency</title>
<meta name="author" content="Chao-Jen Wong">
<meta name="description" content="To obtain the 5’ UTR sequences and their corresponding minimal free energy (MEF) values, we utilized the ShortRead, BioStrings, and BSgenome.Hsapiens.UCSC.hg38 Bioconductor packages, along with...">
<meta name="generator" content="bookdown 0.32 with bs4_book()">
<meta property="og:title" content="Chapter 7 5’ UTR Free Energy Evaluation | DUX4 ribosome footprints profiling and translation efficiency">
<meta property="og:type" content="book">
<meta property="og:url" content="https://FredHutch.github.io/DUX4-IFNg-ribosome-footprints/adjMEF.html">
<meta property="og:description" content="To obtain the 5’ UTR sequences and their corresponding minimal free energy (MEF) values, we utilized the ShortRead, BioStrings, and BSgenome.Hsapiens.UCSC.hg38 Bioconductor packages, along with...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 7 5’ UTR Free Energy Evaluation | DUX4 ribosome footprints profiling and translation efficiency">
<meta name="twitter:description" content="To obtain the 5’ UTR sequences and their corresponding minimal free energy (MEF) values, we utilized the ShortRead, BioStrings, and BSgenome.Hsapiens.UCSC.hg38 Bioconductor packages, along with...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Sen-0.4.5/font.css" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=DM%20Mono&amp;display=swap" rel="stylesheet">
<link href="libs/Bitter-0.4.5/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.2/transition.js"></script><script src="libs/bs3compat-0.4.2/tabs.js"></script><script src="libs/bs3compat-0.4.2/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">DUX4 ribosome footprints profiling and translation efficiency</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="preprocessing.html"><span class="header-section-number">2</span> Preprocessing Ribosome footprints sequencing data</a></li>
<li><a class="" href="rpfs-qc.html"><span class="header-section-number">3</span> Ribosome footprints quality control</a></li>
<li><a class="" href="profiling.html"><span class="header-section-number">4</span> P-sites profiling on genomic features</a></li>
<li><a class="" href="mRNA.html"><span class="header-section-number">5</span> mRNA profiling and DUX4/IFNg induced genes</a></li>
<li><a class="" href="translation-efficiency-CDS.html"><span class="header-section-number">6</span> Translation efficiency changes in CDS</a></li>
<li><a class="active" href="adjMEF.html"><span class="header-section-number">7</span> 5’ UTR Free Energy Evaluation</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints">View book source <i class="fas fa-air-freshener"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="adjMEF" class="section level1" number="7">
<h1>
<span class="header-section-number">7</span> 5’ UTR Free Energy Evaluation<a class="anchor" aria-label="anchor" href="#adjMEF"><i class="fas fa-link"></i></a>
</h1>
<p>To obtain the 5’ UTR sequences and their corresponding minimal free energy (MEF) values, we utilized the <code>ShortRead</code>, <code>BioStrings</code>, and <code>BSgenome.Hsapiens.UCSC.hg38</code> Bioconductor packages, along with the <code>Vienna</code> software. The 5’ UTR models were based on transcripts, but since there could be multiple isoforms with the same 5’ UTR model, we ensured uniqueness by retaining only one instance of each.</p>
<p>In this study, we calculated the free energy for five sets of 5’ UTR features, as follows:</p>
<ol style="list-style-type: decimal">
<li><p><strong>All annotated 5’ UTR</strong>: This set included transcript-based and unique 5’ UTR models extracted from Gencode v3.5, comprising a total of 19,570 genes.</p></li>
<li><p><strong>DUX4 direct targets</strong>: We identified 84 DUX4 directed targets from Polysome-seq data, many of which had alternative promoters identified by <a href="(https://www.nature.com/articles/ng.3846)">Whiddon 2017</a>. These targets were listed in <code>../extdata/DUX4-targets-novel-and-annotated-5'UTR.xlsx</code>. We also checked for alternative splice events, alternative first exon (AFE), or alternative start site (AS) using <code>../scripts/110-HDUX4-promoter.R</code>.</p></li>
<li><p><strong>TE down Ribo-seq</strong>: Down-regulated translational efficiency by Ribo-seq (S6) in 5’UTR region: We identified 1,131 candidates with reduced translational efficiency in the 5’UTR region based on Ribo-seq data.</p></li>
<li><p><strong>Poly-seq up-regulated</strong>: We identified 239 non-DUX4-induced candidates that were up-regulated in DUX4-treatment in Poly-seq samples (High/High).</p></li>
<li><p><strong>Poly-seq down-regulated</strong>: We identified 5,800 candidates that were down-regulated in Poly-seq samples in the High vs. Sub comparison.</p></li>
</ol>
<p>The plan of calculating the the five collections of 5’UTR is implemented by <code>09B-5UTR-structure.R</code>.</p>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:add-5UTR"></span>
<img src="images/boxplot-AdjMEF-5UTR.png" alt="5' UTR minimal free energy per 100 nucleotides" width="675"><p class="caption">
Figure 7.1: 5’ UTR minimal free energy per 100 nucleotides
</p>
</div>

</div>
  <div class="chapter-nav">
<div class="prev"><a href="translation-efficiency-CDS.html"><span class="header-section-number">6</span> Translation efficiency changes in CDS</a></div>
<div class="next"><a href="references.html">References</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav"><li><a class="nav-link" href="#adjMEF"><span class="header-section-number">7</span> 5’ UTR Free Energy Evaluation</a></li></ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints/blob/main/book/08-5UTR-free-energy.Rmd">View source <i class="fas fa-air-freshener"></i></a></li>
          <li><a id="book-edit" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints/edit/main/book/08-5UTR-free-energy.Rmd">Edit this page <i class="fas fa-air-freshener"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>DUX4 ribosome footprints profiling and translation efficiency</strong>" was written by Chao-Jen Wong. It was last built on 2023-04-08.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
