<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 2 Preprocessing Ribosome footprints sequencing data | DUX4 ribosome footprints profiling and translation efficiency</title>
<meta name="author" content="Chao-Jen Wong">
<meta name="description" content="Schema of preprocessing Ribo-seq data This script performs the preprocessing of our ribo-seq data. Below lists the sections of the shell script for each of the preprocessing steps: Software:...">
<meta name="generator" content="bookdown 0.30 with bs4_book()">
<meta property="og:title" content="Chapter 2 Preprocessing Ribosome footprints sequencing data | DUX4 ribosome footprints profiling and translation efficiency">
<meta property="og:type" content="book">
<meta property="og:url" content="https://FredHutch.github.io/DUX4-IFNg-ribosome-footprints/preprocessing.html">
<meta property="og:description" content="Schema of preprocessing Ribo-seq data This script performs the preprocessing of our ribo-seq data. Below lists the sections of the shell script for each of the preprocessing steps: Software:...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 2 Preprocessing Ribosome footprints sequencing data | DUX4 ribosome footprints profiling and translation efficiency">
<meta name="twitter:description" content="Schema of preprocessing Ribo-seq data This script performs the preprocessing of our ribo-seq data. Below lists the sections of the shell script for each of the preprocessing steps: Software:...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Sen-0.4.3/font.css" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=DM%20Mono&amp;display=swap" rel="stylesheet">
<link href="libs/Bitter-0.4.3/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.4.1/transition.js"></script><script src="libs/bs3compat-0.4.1/tabs.js"></script><script src="libs/bs3compat-0.4.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">DUX4 ribosome footprints profiling and translation efficiency</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="active" href="preprocessing.html"><span class="header-section-number">2</span> Preprocessing Ribosome footprints sequencing data</a></li>
<li><a class="" href="rpfs-qc.html"><span class="header-section-number">3</span> Ribosome footprints quality control</a></li>
<li><a class="" href="profiling.html"><span class="header-section-number">4</span> P-sites profiling on genomic features</a></li>
<li><a class="" href="mRNA.html"><span class="header-section-number">5</span> mRNA profiling and DUX4/IFNg induced genes</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints">View book source <i class="fas fa-air-freshener"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="preprocessing" class="section level1" number="2">
<h1>
<span class="header-section-number">2</span> Preprocessing Ribosome footprints sequencing data<a class="anchor" aria-label="anchor" href="#preprocessing"><i class="fas fa-link"></i></a>
</h1>
<div class="inline-figure">
<img src="images/ribo-preprocessing.excalidraw.png" class="cover" width="450" height="500/">Schema of preprocessing Ribo-seq data</div>
<p>This <a href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints/scripts/000-preprocessing_R1.sh">script</a> performs the preprocessing of our ribo-seq data. Below lists the sections of the shell script for each of the preprocessing steps:</p>
<p><strong>Software</strong>:</p>
<ul>
<li>FastQC/0.11.9-Java-11</li>
<li>cutadapt/2.7-GCCcore-8.3.0-Python-3.7.4</li>
<li>Bowtie2/2.4.1-GCCcore-8.3.0</li>
<li>SAMtools/1.10-GCCcore-8.3.0</li>
<li>STAR/2.7.6a-foss-2019b</li>
</ul>
<div id="adapter-trimming" class="section level2" number="2.1">
<h2>
<span class="header-section-number">2.1</span> Adapter trimming<a class="anchor" aria-label="anchor" href="#adapter-trimming"><i class="fas fa-link"></i></a>
</h2>
<p>Trimming forward 3â€™ adapter <code>TGGAATTCTCGGGTGCCAAGG</code> and removing 4 bases from the front and end after trimming:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="preprocessing.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for forward read, trim 3' adapter TGGAATTCTCGGGTGCCAAGG </span></span>
<span id="cb1-2"><a href="preprocessing.html#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">cutadapt</span> <span class="at">--cores</span><span class="op">=</span>4 <span class="at">-O</span> 13 <span class="at">-a</span> N{4}TGGAATTCTCGGGTGCCAAGG <span class="dt">\</span></span>
<span id="cb1-3"><a href="preprocessing.html#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">-u</span> 4 <span class="at">-m</span> 25 <span class="dt">\</span></span>
<span id="cb1-4"><a href="preprocessing.html#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--untrimmed-output</span> <span class="va">${sample_name}</span>_R1_untrimmed.fastq.gz <span class="dt">\ </span></span>
<span id="cb1-5"><a href="preprocessing.html#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="ex">-o</span> <span class="va">${sample_name}</span>_R1_trimmed.fastq.gz <span class="dt">\</span></span>
<span id="cb1-6"><a href="preprocessing.html#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="va">$fq1</span> <span class="op">&gt;</span> <span class="va">${sample_name}</span>_cutadapt_log.txt</span>
<span id="cb1-7"><a href="preprocessing.html#cb1-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb1-8"><a href="preprocessing.html#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="va">fq</span><span class="op">=</span><span class="va">$trim_dir</span>/<span class="va">${sample_name}</span>_R1_trimmed.fastq.gz  </span></code></pre></div>
<p>The <code>-O 13</code> is the minimum overlap for an adapter match, where the 13 is computed as 9 plus 4 (where 9 is the minimum overlap and 4 is the length of the unknown section). If you do not specify it, the adapter sequence would match the end of every read (because N matches anything), and ten bases would then be removed from every read.</p>
<p>This step yielded reads with 25 - 29 nt length.</p>
</div>
<div id="remove-rrna-and-non-rrna-contamination-using-bowtie2" class="section level2" number="2.2">
<h2>
<span class="header-section-number">2.2</span> Remove rRNA and non-rRNA contamination using Bowtie2<a class="anchor" aria-label="anchor" href="#remove-rrna-and-non-rrna-contamination-using-bowtie2"><i class="fas fa-link"></i></a>
</h2>
<p>We first used RNA central database to identify rRNA sequences and built an rRNA reference genome <code>rRNA_ref</code>. Second, we used GFF of Gencode v35 to extract small RNA (tRNA, snoRNA, sRNA, microRNA, etc) sequences and customized a reference genome <code>rsoRNA_ref</code>. We then used <strong>bowtie2</strong> to align the trimmed reads against the rRNA reference genome. Next, we took the un-mapped and aligned against the small RNA reference sequence. The final un-mapped were the reads of interest.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="preprocessing.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rRNA filter</span></span>
<span id="cb2-2"><a href="preprocessing.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2</span> <span class="at">--seed</span><span class="op">=</span>123 <span class="at">--threads</span><span class="op">=</span>32 <span class="at">-x</span> <span class="va">$rRNA_ref</span> <span class="at">-U</span> <span class="va">$fq</span> <span class="at">-S</span> <span class="va">${sample_name}</span>_R1.rRNA.sam <span class="dt">\</span></span>
<span id="cb2-3"><a href="preprocessing.html#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--un-gz</span><span class="op">=</span><span class="va">${sample_name}</span>_R1_rFiltered.fq.gz <span class="dt">\</span></span>
<span id="cb2-4"><a href="preprocessing.html#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="op">&gt;</span> <span class="va">${sample_name}</span>_R1_rRNA.bowtie2.log <span class="dv">2</span><span class="op">&gt;</span> <span class="va">${sample_name}</span>_R1_rRNA.bowtie2.log2</span>
<span id="cb2-5"><a href="preprocessing.html#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># tRNA and snoRNA filter</span></span>
<span id="cb2-6"><a href="preprocessing.html#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">bowtie2</span> <span class="at">--seed</span><span class="op">=</span>123 <span class="at">--threads</span><span class="op">=</span>32 <span class="at">-x</span> <span class="va">$tsoRNA_ref</span> <span class="dt">\</span></span>
<span id="cb2-7"><a href="preprocessing.html#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">-U</span> <span class="va">${sample_name}</span>_R1_rFiltered.fq.gz <span class="dt">\</span></span>
<span id="cb2-8"><a href="preprocessing.html#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">-S</span> <span class="va">${sample_name}</span>_R1.tsoRNA.sam <span class="dt">\</span></span>
<span id="cb2-9"><a href="preprocessing.html#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--un-gz</span><span class="op">=</span><span class="va">${sample_name}</span>_R1_rtsFiltered.fq.gz <span class="dt">\</span></span>
<span id="cb2-10"><a href="preprocessing.html#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="op">&gt;</span> <span class="va">${sample_name}</span>_R1_tsoRNA.bowtie2.log <span class="dv">2</span><span class="op">&gt;</span>   <span class="va">${sample_name}</span>_R1_tsoRNA.bowtie2.log2</span></code></pre></div>
</div>
<div id="alignment" class="section level2" number="2.3">
<h2>
<span class="header-section-number">2.3</span> Alignment<a class="anchor" aria-label="anchor" href="#alignment"><i class="fas fa-link"></i></a>
</h2>
<p>We took the final un-mapped reads from previous step and aligned against GRCh38.p13 by <strong>STAR</strong>:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="preprocessing.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">STAR</span> <span class="at">--runThreadN</span> 4 <span class="at">--outSAMattributes</span> NH HI NM MD AS nM jM jI XS <span class="dt">\</span></span>
<span id="cb3-2"><a href="preprocessing.html#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">--genomeDir</span> <span class="va">$star_index</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="preprocessing.html#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">--readFilesIn</span> <span class="va">$filtered_dir</span>/<span class="va">${sample_name}</span>_R1_rtsFiltered.fq.gz <span class="dt">\</span></span>
<span id="cb3-4"><a href="preprocessing.html#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFileNamePrefix</span> <span class="va">$bam_dir</span>/<span class="va">${sample_name}</span>_ <span class="dt">\</span></span>
<span id="cb3-5"><a href="preprocessing.html#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">--sjdbOverhang</span> 25 <span class="dt">\</span></span>
<span id="cb3-6"><a href="preprocessing.html#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFilterScoreMinOverLread</span> 0 <span class="dt">\ </span></span>
<span id="cb3-7"><a href="preprocessing.html#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="ex">--outFilterMatchNminOverLread</span> 0 <span class="dt">\</span></span>
<span id="cb3-8"><a href="preprocessing.html#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outFilterMatchNmin</span> 20 <span class="dt">\</span></span>
<span id="cb3-9"><a href="preprocessing.html#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">--readFilesCommand</span> zcat <span class="dt">\</span></span>
<span id="cb3-10"><a href="preprocessing.html#cb3-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">--outSAMtype</span> BAM SortedByCoordinate</span>
<span id="cb3-11"><a href="preprocessing.html#cb3-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-12"><a href="preprocessing.html#cb3-12" aria-hidden="true" tabindex="-1"></a><span class="ex">samtools</span> index <span class="va">$bam_dir</span>/<span class="va">${sample_name}</span>_Aligned.sortedByCoord.out.bam</span></code></pre></div>
<p>See next chapter for Ribo-seq specific quality control.</p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="index.html"><span class="header-section-number">1</span> About</a></div>
<div class="next"><a href="rpfs-qc.html"><span class="header-section-number">3</span> Ribosome footprints quality control</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#preprocessing"><span class="header-section-number">2</span> Preprocessing Ribosome footprints sequencing data</a></li>
<li><a class="nav-link" href="#adapter-trimming"><span class="header-section-number">2.1</span> Adapter trimming</a></li>
<li><a class="nav-link" href="#remove-rrna-and-non-rrna-contamination-using-bowtie2"><span class="header-section-number">2.2</span> Remove rRNA and non-rRNA contamination using Bowtie2</a></li>
<li><a class="nav-link" href="#alignment"><span class="header-section-number">2.3</span> Alignment</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints/blob/main/book/01-preprocessiong-ribo-seq.Rmd">View source <i class="fas fa-air-freshener"></i></a></li>
          <li><a id="book-edit" href="https://github.com/FredHutch/DUX4-IFNg-ribosome-footprints/edit/main/book/01-preprocessiong-ribo-seq.Rmd">Edit this page <i class="fas fa-air-freshener"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>DUX4 ribosome footprints profiling and translation efficiency</strong>" was written by Chao-Jen Wong. It was last built on 2022-11-30.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
